### DISCUSSION TITLE
Coupling of Lagrangian particles and continuous phase grids

### DISCUSSION THREAD

**Discussion by klrqst** [AMReX-Combustion/PeleLMeX#375] [Q&A]
Dear AMReX-Combustion community,

I am writing to seek guidance on a issue I am facing while using PeleLMeX and PeleMP for my simulations. As you know, PeleMP employs the PCM (Particle Centroid Method) to couple the Lagrangian and Eulerian terms effectively. However, I have encountered a challenge when working with large-diameter particles.

The problem arises when the ratio of particle diameter to the continuous phase grid size approaches or exceeds 1. In such cases, the error associated with the coupling becomes unpredictable. To address this, I am considering a modification to distribute the source term of the Lagrangian particles across multiple grids, instead of concentrating it in a single grid cell.

I would greatly appreciate it if you could kindly offer some suggestions on how I might incorporate this concept into PeleLMeX and PeleMP. Thank you in advance for your support and assistance.

Best regards

**Comment by baperry2** [comment]
Note that PeleMP is no longer developed as a separate library and has been folded into PelePhysics. 

Droplet diameters exceeding the grid size are not the [intended use case](https://amrex-combustion.github.io/PelePhysics/Spray.html) of the Lagrangian spray particle capability, so it may be worth evaluating whether this is the appropriate tool for you. If you do want to distribute the source term across multiple cells, one potential option would be to apply the Filter utility form PelePhysics to filter the spray source terms before adding them to the rest of the sources in PeleLMeX, although you might need to think carefully about how to do this in a consistent way for e.g. multilevel simulations.

**Reply by klrqst** [reply]
Thank you very much for your response. I see that the spray source term is added in `SprayParticles.cpp`, and `ijkc` represents the cell corresponding to the particle. I would like to use the Filter utility to also add the grids around `ijkc` to the source term. Could you please provide some more guidance? Thank you again.

**Reply by baperry2** [reply]
The filter utility works on FABs not point data. Just spitballing but I'd imagine you'd apply the filter utility to the data structure containing the spray source term in the `SprayAddSource` function in `PeleLMeX_SprayParticles.cpp` prior to it being added to the rest of the source terms. For a relatively simple example of how the filter utility works, check out the filterPlt tool from PeleAnalysis: https://github.com/AMReX-Combustion/PeleAnalysis/blob/master/Src/filterPlt.cpp#L217. Note the filters are not EB aware.

**Reply by klrqst** [reply]
Thank you very much for your response, but I encountered some issues in the application, sorry to disturb you. I added these two lines of code at the end of the` SprayAddSource` function:
```
Filter filter(les_filter_type, les_filter_fgr);  
filter.apply_filter(extsource, extsource);
```
When `les_filter_type=1` and `les_filter_fgr=1`, the sample program can run, but it seems that the filter doesn't have any effect. When `les_filter_type` or `les_filter_fgr` is set to other values, the sample program will report an error. The error message is as follows:

>  ====================   NEW TIME STEP   ==================== 
>  STEP [0] - Time: 0, dt 1e-08
> moveKickDrift ... updating particle positions and velocity
> SprayMKDLevel 0
> MKD on Active particles on level 0
>    SDC iter [1] 
> MLMG: # of AMR levels: 1
>       # of MG levels on the coarsest AMR level: 7
> MLMG: Initial rhs               = 4.26660857e+263
> MLMG: Initial residual (resid0) = 4.26660857e+263
> MLMG: Iteration   1 Fine resid/bnorm = 0
> MLMG: Final Iter. 1 resid, resid/bnorm = 0, 0
> MLMG: Timers: Solve = 0.000816137 Iter = 0.000584609 Bottom = 1.6493e-05
>    - oneSDC()::MACProjection()   --> Time: 0.003555749077
>    - oneSDC()::ScalarAdvection() --> Time: 0.01018678816
>    - oneSDC()::ScalarDiffusion() --> Time: 0.02846485283
> From CVODE: At t = 0 and h = 1.3093e-18, the corrector convergence test failed repeatedly or with |h| = hmin.
> From CVODE: At t = 0 and h = 1.3093e-18, the corrector convergence test failed repeatedly or with |h| = hmin.
> From CVODE: At t = 0 and h = 1.3093e-18, the corrector convergence test failed repeatedly or with |h| = hmin.
> amrex::Abort::33::Aborting from CVODE !!!
> SIGABRT
> amrex::Abort::38::Aborting from CVODE !!!
> SIGABRT
> amrex::Abort::7::Aborting from CVODE !!!
> SIGABRT
> From CVODE: At t = 0 and h = 1.3093e-18, the corrector convergence test failed repeatedly or with |h| = hmin.
> amrex::Abort::13::Aborting from CVODE !!!
> SIGABRT
> From CVODE: At t = 0 and h = 1.3093e-18, the corrector convergence test failed repeatedly or with |h| = hmin.
> amrex::Abort::5::Aborting from CVODE !!!
> SIGABRT
> From CVODE: At t = 0 and h = 1.3093e-18, the corrector convergence test failed repeatedly or with |h| = hmin.
> amrex::Abort::34::Aborting from CVODE !!!
> SIGABRT
> From CVODE: At t = 0 and h = 1.3093e-18, the corrector convergence test failed repeatedly or with |h| = hmin.
> From CVODE: At t = 0 and h = 1.3093e-18, the corrector convergence test failed repeatedly or with |h| = hmin.
> amrex::Abort::26::Aborting from CVODE !!!
> SIGABRT
> amrex::Abort::24::Aborting from CVODE !!!
> SIGABRT
> amrex::Abort::0::Aborting from CVODE !!!
> SIGABRT
> From CVODE: At t = 0 and h = 1.3093e-18, the corrector convergence test failed repeatedly or with |h| = hmin.
> From CVODE: At t = 0 and h = 1.3093e-18, the corrector convergence test failed repeatedly or with |h| = hmin.
> amrex::Abort::16::Aborting from CVODE !!!
> SIGABRT
> From CVODE: At t = 0 and h = 1.3093e-18, the corrector convergence test failed repeatedly or with |h| = hmin.
> From CVODE: At t = 0 and h = 1.3093e-18, the corrector convergence test failed repeatedly or with |h| = hmin.
> From CVODE: At t = 0 and h = 1.3093e-18, the corrector convergence test failed repeatedly or with |h| = hmin.
> amrex::Abort::39::Aborting from CVODE !!!
> SIGABRT
> amrex::Abort::20::Aborting from CVODE !!!
> SIGABRT
> amrex::Abort::3::Aborting from CVODE !!!
> SIGABRT
> From CVODE: At t = 0 and h = 1.3093e-18, the corrector convergence test failed repeatedly or with |h| = hmin.
> From CVODE: At t = 0 and h = 1.3093e-18, the corrector convergence test failed repeatedly or with |h| = hmin.
> amrex::Abort::18::Aborting from CVODE !!!
> SIGABRT
> amrex::Abort::32::Aborting from CVODE !!!
> SIGABRT
> --------------------------------------------------------------------------
> 

Could you please advise what the reason might be? Thank you again.

**Reply by baperry2** [reply]
Often these CVODE failures are a symptom of failing upstream of CVODE resulting in a bad state getting passed to CVODE. You're trying to do the filter operation in place, which the utility is not intended for. The output must go into a new FAB. Also make sure you have enough ghost cells for the filter stencil you're using.
